### PR TITLE
feat: coinbase wallet

### PR DESCRIPTION
**Pull request by asamere**

### CODE REVIEW

File: packages/hdwallet-coinbase/src/adapter.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  public async initialize(): Promise<number> {
+    return Object.keys(this.keyring.wallets).length;
+  }
+
+  public async pairDevice(): Promise<CoinbaseHDWallet | undefined> {
+    console.info("coinbase-wallet: pairDevice");
+    const APP_NAME = "ShapeShift";
+    const APP_LOGO_URL = "https://avatars.githubusercontent.com/u/52928763?s=50&v=4";
+    const DEFAULT_ETH_JSONRPC_URL = `https://mainnet.infura.io/v3/6e2f28ff4f5340fdb0db5da3baec0af2`;
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/adapter.ts:25]: can we point this at `https://daemon.ethereum.shapeshift.com`?

File: packages/hdwallet-coinbase/src/adapter.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      appName: APP_NAME,
+      appLogoUrl: APP_LOGO_URL,
+      darkMode: false,
+    });
+
+    // Initialize a Web3 Provider object
+    const coinbaseWalletProvider = coinbaseWallet.makeWeb3Provider(DEFAULT_ETH_JSONRPC_URL, DEFAULT_CHAIN_ID);
+
+    // Initialize a Web3 object
+    // const web3 = new Web3(coinbaseWalletProvider as any);
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/adapter.ts:39]: remove?

File: packages/hdwallet-coinbase/src/adapter.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+    const wallet = new CoinbaseHDWallet(coinbaseWalletProvider);
+    await wallet.initialize();
+    const deviceID = await wallet.getDeviceID();
+    console.info(`deviceID: ${deviceID}`);
+
+    this.keyring.add(wallet, deviceID);
+    this.keyring.emit(["Coinbase", deviceID, core.Events.CONNECT], deviceID);
+    // Optionally, have the default account set for web3.js
+    // web3.eth.defaultAccount = accounts[0]
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/adapter.ts:57]: remove?

File: packages/hdwallet-coinbase/src/coinbase.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+  }
+
+  public supportsBroadcast(): boolean {
+    return true;
+  }
+
+  public describePath(msg: core.DescribePath): core.PathDescription {
+    switch (msg.coin) {
+      case "Ethereum":
+        // return eth.describeETHPath(msg.path);
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/coinbase.ts:61]: delet?

File: packages/hdwallet-coinbase/src/coinbase.ts [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+        return core.describeETHPath(msg.path);
+      default:
+        throw new Error("Unsupported path");
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  public ethNextAccountPath(msg: core.ETHAccountPath): core.ETHAccountPath | undefined {
+    // TODO: What do we do here?
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/coinbase.ts:70]: same as whatever metamask does

File: packages/hdwallet-coinbase/src/coinbase.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    return undefined;
+  }
+
+  public async ethSupportsNetwork(chainId: number): Promise<boolean> {
+    return chainId === 1;
```
Comment (0xdef1cafe) [inline packages/hdwallet-coinbase/src/coinbase.ts:75]: https://help.coinbase.com/en/wallet/getting-started/what-types-of-crypto-does-wallet-support

looks like all our evm chains should be supported

File: packages/hdwallet-coinbase/src/coinbase.ts (hunk 5 above)
Comment (asamere) [reply packages/hdwallet-coinbase/src/coinbase.ts:70]: metamask asks "What do we do here?"

File: packages/hdwallet-coinbase/src/coinbase.ts (hunk 6 above)
Comment (asamere) [reply packages/hdwallet-coinbase/src/coinbase.ts:75]: no idea what the term "eth supports network" is supposed to mean but this is also metamask verbatim

File: packages/hdwallet-coinbase/src/adapter.ts (hunk 1 above)
Comment (asamere) [reply packages/hdwallet-coinbase/src/adapter.ts]: haha crossed paths. added CoinbaseProviderConfig and pass from web (borrowed from WalletConnect). Should be able to point it anywhere you like via `.env`.

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**